File: java/0012-integer-to-roman.java
import java.util.*;

/**
 * LeetCode Problem 0012: Integer to Roman
 *
 * Problem:
 * Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.
 *
 * Symbol       Value
 * I             1
 * V             5
 * X             10
 * L             50
 * C             100
 * D             500
 * M             1000
 *
 * Given an integer, convert it to a roman numeral. You may assume the input is always between 1 and 3999.
 *
 * Examples:
 * 1. Input: 3, Output: "III"
 * 2. Input: 58, Output: "LVIII" (50 + 5 + 3 = LVIII)
 * 3. Input: 1994, Output: "MCMXCIV" (1000 + 900 + 90 + 4 = MCMXCIV)
 *
 * Constraints:
 * - 1 <= num <= 3999
 */

public class IntegerToRoman {

    // Q1: Greedy approach using value-symbol pairs in descending order
    public static String intToRomanGreedy(int num) {
        /**
         * Time Complexity: O(1) - maximum iterations is bounded (num can be at most 3999)
         * Space Complexity: O(1) - output string is at most 12 characters
         *
         * Approach: Start from largest values and greedily build the roman numeral string.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            int count = num / values[i];
            if (count > 0) {
                for (int j = 0; j < count; j++) {
                    result.append(symbols[i]);
                }
                num -= values[i] * count;
            }
        }
        return result.toString();
    }

    // Q2: Using HashMap for value-symbol mapping
    public static String intToRomanHashMap(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Using LinkedHashMap to preserve insertion order of value-symbol pairs.
         */
        LinkedHashMap<Integer, String> valToSym = new LinkedHashMap<>();
        valToSym.put(1000, "M");
        valToSym.put(900, "CM");
        valToSym.put(500, "D");
        valToSym.put(400, "CD");
        valToSym.put(100, "C");
        valToSym.put(90, "XC");
        valToSym.put(50, "L");
        valToSym.put(40, "XL");
        valToSym.put(10, "X");
        valToSym.put(9, "IX");
        valToSym.put(5, "V");
        valToSym.put(4, "IV");
        valToSym.put(1, "I");

        StringBuilder result = new StringBuilder();
        for (Map.Entry<Integer, String> entry : valToSym.entrySet()) {
            int value = entry.getKey();
            String symbol = entry.getValue();
            while (num >= value) {
                result.append(symbol);
                num -= value;
            }
        }
        return result.toString();
    }

    // Q3: Using while loop with single loop
    public static String intToRomanSingleLoop(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Single loop with pointer advancement for cleaner code.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();
        int index = 0;

        while (num > 0) {
            if (num >= values[index]) {
                result.append(symbols[index]);
                num -= values[index];
            } else {
                index++;
            }
        }
        return result.toString();
    }

    // Q4: Using repeat method for cleaner append
    public static String intToRomanRepeat(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Use String.repeat() for multiple appends (Java 11+).
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            int count = num / values[i];
            if (count > 0) {
                result.append(symbols[i].repeat(count));
                num -= values[i] * count;
            }
        }
        return result.toString();
    }

    // Q5: Using arrays of digit mappings for each place value
    public static String intToRomanGrouped(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Group the values by place value (thousands, hundreds, tens, ones).
         */
        String[] thousands = {"", "M", "MM", "MMM"};
        String[] hundreds = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
        String[] tens = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
        String[] ones = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};

        return thousands[num / 1000] +
               hundreds[(num % 1000) / 100] +
               tens[(num % 100) / 10] +
               ones[num % 10];
    }

    // Q6: Using helper method for repeated symbol appending
    public static String intToRomanHelper(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Extract symbol appending logic into a helper method.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            result.append(repeatSymbol(symbols[i], num / values[i]));
            num %= values[i];
        }
        return result.toString();
    }

    private static String repeatSymbol(String symbol, int count) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < count; i++) {
            sb.append(symbol);
        }
        return sb.toString();
    }

    // Q7: Using array of Pair objects for cleaner iteration
    public static String intToRomanPair(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Use a list of pairs for more organized data structure.
         */
        List<int[]> pairs = Arrays.asList(
            new int[]{1000, 'M'}, new int[]{900, 'C'}, new int[]{500, 'D'}, new int[]{400, 'C'},
            new int[]{100, 'C'}, new int[]{90, 'X'}, new int[]{50, 'L'}, new int[]{40, 'X'},
            new int[]{10, 'X'}, new int[]{9, 'I'}, new int[]{5, 'V'}, new int[]{4, 'I'}, new int[]{1, 'I'}
        );

        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            while (num >= values[i]) {
                result.append(symbols[i]);
                num -= values[i];
            }
        }
        return result.toString();
    }

    // Q8: Using do-while with explicit index advancement
    public static String intToRomanDoWhile(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Alternative loop structure using do-while.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();
        int index = 0;

        do {
            if (num >= values[index]) {
                result.append(symbols[index]);
                num -= values[index];
            } else {
                index++;
            }
        } while (num > 0 && index < values.length);

        return result.toString();
    }

    // Q9: Using recursive approach
    public static String intToRomanRecursive(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1) - recursion depth is bounded
         *
         * Convert by recursively handling the largest value first.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        return intToRomanRecursiveHelper(num, 0, new StringBuilder(), values, symbols).toString();
    }

    private static StringBuilder intToRomanRecursiveHelper(int num, int index, StringBuilder result,
                                                           int[] values, String[] symbols) {
        if (num == 0 || index >= values.length) {
            return result;
        }

        if (num >= values[index]) {
            result.append(symbols[index]);
            return intToRomanRecursiveHelper(num - values[index], index, result, values, symbols);
        } else {
            return intToRomanRecursiveHelper(num, index + 1, result, values, symbols);
        }
    }

    // Q10: Optimized with early exit
    public static String intToRomanOptimized(int num) {
        /**
         * Time Complexity: O(1)
         * Space Complexity: O(1)
         *
         * Optimized version with early exit condition when num becomes 0.
         */
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder result = new StringBuilder();

        for (int i = 0; i < values.length && num > 0; i++) {
            while (num >= values[i]) {
                result.append(symbols[i]);
                num -= values[i];
            }
        }

        return result.toString();
    }

    public static void main(String[] args) {
        int[][] testCases = {
            {3, 0},           // III
            {58, 1},          // LVIII
            {1994, 2},        // MCMXCIV
            {1, 3},           // I
            {9, 4},           // IX
            {27, 5},          // XXVII
            {49, 6},          // XLIX
            {3999, 7},        // MMMCMXCIX
            {444, 8},         // CDXLIV
            {2023, 9}         // MMXXIII
        };

        String[] expectedResults = {
            "III", "LVIII", "MCMXCIV", "I", "IX",
            "XXVII", "XLIX", "MMMCMXCIX", "CDXLIV", "MMXXIII"
        };

        java.lang.reflect.Method[] methods = IntegerToRoman.class.getDeclaredMethods();
        java.util.Arrays.sort(methods, (a, b) -> a.getName().compareTo(b.getName()));

        int questionNum = 1;
        for (java.lang.reflect.Method method : methods) {
            if (!method.getName().startsWith("intToRoman") ||
                method.getName().contains("Helper") ||
                method.getName().contains("Recursive")) {
                continue;
            }

            System.out.println("\nQ" + questionNum + ": " + method.getName());
            System.out.println("-".repeat(50));

            boolean allPassed = true;
            for (int i = 0; i < testCases.length; i++) {
                int num = testCases[i][0];
                String expected = expectedResults[i];

                try {
                    String result = (String) method.invoke(null, num);
                    boolean passed = result.equals(expected);
                    allPassed = allPassed && passed;
                    String status = passed ? "✓" : "✗";
                    System.out.println("  " + status + " intToRoman(" + num + ") = " + result +
                                     " (expected: " + expected + ")");
                } catch (Exception e) {
                    System.out.println("  ✗ Error testing " + method.getName());
                    allPassed = false;
                }
            }

            if (allPassed) {
                System.out.println("  All tests passed!");
            }
            questionNum++;
        }

        // Special testing for recursive method
        System.out.println("\nQ9: intToRomanRecursive");
        System.out.println("-".repeat(50));
        boolean recursivePassed = true;
        for (int i = 0; i < testCases.length; i++) {
            int num = testCases[i][0];
            String expected = expectedResults[i];
            String result = intToRomanRecursive(num);
            boolean passed = result.equals(expected);
            recursivePassed = recursivePassed && passed;
            String status = passed ? "✓" : "✗";
            System.out.println("  " + status + " intToRoman(" + num + ") = " + result +
                             " (expected: " + expected + ")");
        }
        if (recursivePassed) {
            System.out.println("  All tests passed!");
        }
    }
}

File: python/0012-integer-to-roman.py
"""
LeetCode Problem 0012: Integer to Roman

Problem:
Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.

Symbol       Value
I             1
V             5
X             10
L             50
C             100
D             500
M             1000

Given an integer, convert it to a roman numeral. You may assume the input is always between 1 and 3999.

Examples:
1. Input: 3, Output: "III"
2. Input: 58, Output: "LVIII" (50 + 5 + 3 = LVIII)
3. Input: 1994, Output: "MCMXCIV" (1000 + 900 + 90 + 4 = MCMXCIV)

Constraints:
- 1 <= num <= 3999
"""

# Q1: Greedy approach using value-symbol pairs in descending order
def int_to_roman_greedy(num: int) -> str:
    """
    Time Complexity: O(1) - maximum iterations is bounded (num can be at most 3999)
    Space Complexity: O(1) - output string is at most 12 characters

    Approach: Start from largest values and greedily build the roman numeral string.
    """
    values = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = ""
    for value, symbol in values:
        count = num // value
        if count:
            result += symbol * count
            num -= value * count

    return result


# Q2: Same approach with different data structure (dictionary)
def int_to_roman_dict(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Using a dictionary for value-symbol mapping with explicit ordering.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = []
    for value, symbol in val_sym:
        while num >= value:
            result.append(symbol)
            num -= value

    return ''.join(result)


# Q3: Recursive approach
def int_to_roman_recursive(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1) - recursion depth is bounded

    Convert by recursively handling the largest value first.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    def helper(n, index):
        if n == 0:
            return ""

        value, symbol = val_sym[index]
        if n >= value:
            return symbol + helper(n - value, index)
        else:
            return helper(n, index + 1)

    return helper(num, 0)


# Q4: Using divmod for cleaner code
def int_to_roman_divmod(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Use divmod to separate quotient and remainder in one operation.
    """
    values = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = ""
    for value, symbol in values:
        count, num = divmod(num, value)
        result += symbol * count

    return result


# Q5: Subtractive notation grouping
def int_to_roman_grouped(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Group the values by place value (thousands, hundreds, tens, ones).
    """
    thousands = ['', 'M', 'MM', 'MMM']
    hundreds = ['', 'C', 'CC', 'CCC', 'CD', 'D', 'DC', 'DCC', 'DCC', 'CM']
    tens = ['', 'X', 'XX', 'XXX', 'XL', 'L', 'LX', 'LXX', 'LXX', 'XC']
    ones = ['', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX']

    return (thousands[num // 1000] +
            hundreds[(num % 1000) // 100] +
            tens[(num % 100) // 10] +
            ones[num % 10])


# Q6: Iterative with list building
def int_to_roman_list(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Build result list and join at the end for efficiency.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = []
    for value, symbol in val_sym:
        count = num // value
        result.extend([symbol] * count)
        num %= value

    return ''.join(result)


# Q7: Generator approach
def int_to_roman_generator(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Use a generator for efficient symbol production.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    def gen_symbols():
        for value, symbol in val_sym:
            for _ in range(num // value):
                yield symbol
            num_ref[0] -= (num // value) * value

    num_ref = [num]
    return ''.join(gen_symbols())


# Q8: While loop with explicit pointer
def int_to_roman_pointer(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Maintain a pointer to current position in value-symbol pairs.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = ""
    i = 0
    while num > 0 and i < len(val_sym):
        value, symbol = val_sym[i]
        if num >= value:
            result += symbol
            num -= value
        else:
            i += 1

    return result


# Q9: Two-pointer approach
def int_to_roman_two_pointer(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Use two pointers - one for tracking value-symbol pair, one for string building.
    """
    val_sym = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'),
        (1, 'I')
    ]

    result = []
    index = 0

    while num > 0:
        if num >= val_sym[index][0]:
            result.append(val_sym[index][1])
            num -= val_sym[index][0]
        else:
            index += 1

    return ''.join(result)


# Q10: Hardcoded strings for each digit position
def int_to_roman_hardcoded(num: int) -> str:
    """
    Time Complexity: O(1)
    Space Complexity: O(1)

    Pre-computed strings for all possible digit combinations.
    """
    ones = ['', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX']
    tens = ['', 'X', 'XX', 'XXX', 'XL', 'L', 'LX', 'LXX', 'LXXX', 'XC']
    hundreds = ['', 'C', 'CC', 'CCC', 'CD', 'D', 'DC', 'DCC', 'DCCC', 'CM']
    thousands = ['', 'M', 'MM', 'MMM']

    return (thousands[num // 1000] +
            hundreds[(num % 1000) // 100] +
            tens[(num % 100) // 10] +
            ones[num % 10])


if __name__ == "__main__":
    test_cases = [
        (3, "III"),
        (58, "LVIII"),
        (1994, "MCMXCIV"),
        (1, "I"),
        (9, "IX"),
        (27, "XXVII"),
        (49, "XLIX"),
        (3999, "MMMCMXCIX"),
        (444, "CDXLIV"),
        (2023, "MMXXIII"),
    ]

    functions = [
        int_to_roman_greedy,
        int_to_roman_dict,
        int_to_roman_recursive,
        int_to_roman_divmod,
        int_to_roman_grouped,
        int_to_roman_list,
        int_to_roman_generator,
        int_to_roman_pointer,
        int_to_roman_two_pointer,
        int_to_roman_hardcoded,
    ]

    for i, func in enumerate(functions, 1):
        print(f"\nQ{i}: {func.__name__}")
        print("-" * 50)
        all_passed = True
        for num, expected in test_cases:
            result = func(num)
            passed = result == expected
            all_passed = all_passed and passed
            status = "✓" if passed else "✗"
            print(f"  {status} int_to_roman({num}) = {result} (expected: {expected})")

        if all_passed:
            print(f"  All tests passed!")
